DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc

```diff
@@ -0,0 +1,201 @@
+
+  int RPindexValid[RPotsTotalNumber];
+  MonitorElement *h2HitsVsBXRandoms[RPotsTotalNumber];
+
+  int getRPindex(int arm, int station, int rp) {
+    if (arm < 0 || station < 0 || rp < 0)
+      return (IndexNotValid);
+    if (arm > 1 || station >= NStationMAX || rp >= NRPotsMAX)
+      return (IndexNotValid);
+    int rc = (arm * NStationMAX + station) * NRPotsMAX + rp;
+    return (rc);
+  }
+};
+
+//----------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
+
+//-------------------------------------------------------------------------------
+
+CTPPSRandomDQMSource::CTPPSRandomDQMSource(const edm::ParameterSet &ps)
+    : rpStatusWord(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
+  tokenDigi = consumes<DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"));
+}
+
+//--------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::dqmBeginRun(edm::Run const &run, edm::EventSetup const &) {
+  for (int stn = 0; stn < StationIDMAX; stn++) {
+    StationStatus[stn] = 0;
+    for (int rp = 0; rp < RPotsIDMAX; rp++)
+      RPstatus[stn][rp] = 0;
+  }
+
+  unsigned int rpSts = rpStatusWord << 1;
+  for (int stn = 0; stn < NStationMAX; stn++) {
+    int stns = 0;
+    for (int rp = 0; rp < NRPotsMAX; rp++) {
+      rpSts = (rpSts >> 1);
+      RPstatus[stn][rp] = rpSts & 1;
+      if (RPstatus[stn][rp] > 0)
+        stns = 1;
+    }
+    StationStatus[stn] = stns;
+  }
+
+  for (int index = 0; index < 2 * 3 * NRPotsMAX; index++)
+    RPindexValid[index] = 0;
+}
+
+//-------------------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::bookHistograms(DQMStore::IBooker &ibooker, edm::Run const &, edm::EventSetup const &) {
+  ibooker.cd();
+  ibooker.setCurrentFolder("CTPPS/TrackingPixel");
+
+  hBX = ibooker.book1D("events per BX", "ctpps_pixel;Event.BX", 4002, -1.5, 4000. + 0.5);
+
+  for (int arm = 0; arm < NArms; arm++) {
+    CTPPSDetId ID(CTPPSDetId::sdTrackingPixel, arm, 0);
+    string sd;
+    ID.armName(sd, CTPPSDetId::nPath);
+
+    ibooker.setCurrentFolder(sd);
+
+    for (int stn = 0; stn < NStationMAX; stn++) {
+      if (StationStatus[stn] == 0)
+        continue;
+      ID.setStation(stn);
+      string stnd;
+      CTPPSDetId(ID.stationId()).stationName(stnd, CTPPSDetId::nPath);
+
+      ibooker.setCurrentFolder(stnd);
+
+      for (int rp = RPn_first; rp < RPn_last; rp++) {  // only installed pixel pots
+        ID.setRP(rp);
+        string rpd, rpTitle;
+        CTPPSDetId(ID.rpId()).rpName(rpTitle, CTPPSDetId::nFull);
+        CTPPSDetId(ID.rpId()).rpName(rpd, CTPPSDetId::nPath);
+
+        ibooker.setCurrentFolder(rpd);
+
+        int indexP = getRPindex(arm, stn, rp);
+        RPindexValid[indexP] = 1;
+
+        h2HitsVsBXRandoms[indexP] = ibooker.book2D("Digi per plane per BX - random triggers",
```

Comment:
Is `indexP` guaranteed to be inbounds? (i.e. smaller than the size of `h2HitsVsBXRandoms`)

Reply:
Yes, getRPindex allows for a maximum return value of 35, while kRPotsTotalNumber_ (which is the size of the mentioned arrays) is equal to 36. The two variables are linked, and the maximum adjusts depending on the other const variables which act as settings.